lp2p/peer.go

```diff
@@ -0,0 +1,386 @@
+}
+
+func (p *PeerSet) GetByID(id peer.ID) p2p.Peer {
+	peer, err := p.getOrAdd(id)
+	if err != nil {
+		p.logger.Error("PeerSet.GetByID failed", "peer_id", id.String(), "err", err)
+		return nil
+	}
+
+	return peer
+}
+
+func (p *PeerSet) Get(key p2p.ID) p2p.Peer {
+	id := p.keyToPeerID(key)
+	if id == "" {
+		return nil
+	}
+
+	return p.GetByID(id)
+}
+
+func (p *PeerSet) getOrAdd(id peer.ID) (*Peer, error) {
+	// use cache
+	if peer, ok := p.cacheGet(id); ok {
+		return peer, nil
+	}
+
+	// we don't want to return self
+	if p.host.ID() == id {
+		return nil, nil
+	}
+
+	addrInfo := p.host.Peerstore().PeerInfo(id)
+	if len(addrInfo.Addrs) == 0 {
+		return nil, errors.New("peer has no addresses in peerstore")
+	}
+
+	peer, err := NewPeer(p.host, addrInfo, p.metrics)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create peer")
+	}
+
+	peer.SetLogger(p.logger.With("peer_id", id.String()))
+
+	return p.cacheSet(peer), nil
+}
+
+func (p *PeerSet) Remove(key p2p.ID) {
+	if id := p.keyToPeerID(key); id != "" {
+		p.remove(id)
+	}
+}
+
+func (p *PeerSet) remove(id peer.ID) {
+	if _, ok := p.cacheGet(id); !ok {
+		// noop
+		return
+	}
+
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	// delete *Peer
+	delete(p.peers, id)
+
+	// drop kv if any
+	p.host.Peerstore().RemovePeer(id)
+}
+
+func (p *PeerSet) Size() int {
+	return len(p.existingPeerIDs())
+}
+
+func (p *PeerSet) Copy() []p2p.Peer {
+	peers := p.existingPeerIDs()
+	results := make([]p2p.Peer, 0, len(peers))
+
+	for _, id := range peers {
+		key := peerIDToKey(id)
+
+		if peer := p.Get(key); peer != nil {
+			results = append(results, peer)
+		}
+	}
+
+	return results
+}
+
+func (p *PeerSet) ForEach(lambda func(p2p.Peer)) {
+	peers := p.existingPeerIDs()
+
+	for _, id := range peers {
+		key := peerIDToKey(id)
+		peer := p.Get(key)
+
+		if peer == nil {
+			continue
+		}
+
+		lambda(peer)
+	}
+}
+
+func (p *PeerSet) Random() p2p.Peer { return nil }
+
+func (p *PeerSet) existingPeerIDs() []peer.ID {
+	hostID := p.host.ID()
+	peers := p.host.Peerstore().PeersWithAddrs()
+
+	// exclude self
+	for i := 0; i < len(peers); i++ {
+		if peers[i] == hostID {
+			peers = append(peers[:i], peers[i+1:]...)
+			break
+		}
+	}
+
+	p.logger.Debug("Existing peer IDs", "host_id", hostID, "peers", peers)
+
+	return peers
+}
+
+func (p *PeerSet) cacheGet(id peer.ID) (*Peer, bool) {
+	p.mu.RLock()
+	defer p.mu.RUnlock()
+
+	peer, ok := p.peers[id]
+
+	return peer, ok
+}
+
+func (p *PeerSet) cacheSet(peer *Peer) *Peer {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	// noop
+	if peer, ok := p.peers[peer.addrInfo.ID]; ok {
+		return peer
+	}
+
+	p.peers[peer.addrInfo.ID] = peer
+
+	return peer
+}
+
+func (p *PeerSet) keyToPeerID(key p2p.ID) peer.ID {
+	if key == "" {
+		return ""
+	}
+
+	id, err := keyToPeerID(key)
+	if err != nil {
+		p.logger.Error("Failed to convert key to peer ID!", "peer_key", key, "err", err)
+		return ""
+	}
+
+	return id
+}
+
+func keyToPeerID(key p2p.ID) (peer.ID, error) {
```

Comment:
nit: does this need to be a separate function? It's only used within the peer set method and that's the only thing called by other peer set methods.